temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java

```diff
@@ -100,21 +106,162 @@ public <T> T fromPayloads(
       int index, Optional<Payloads> content, Class<T> valueType, Type valueGenericType)
       throws DataConverterException {
     if (content.isPresent()) {
-      List<Payload> decodedPayloads = codec.decode(content.get().getPayloadsList());
-      content = Optional.of(Payloads.newBuilder().addAllPayloads(decodedPayloads).build());
+      content = Optional.of(decode(content.get()));
     }
     return dataConverter.fromPayloads(index, content, valueType, valueGenericType);
   }
 
+  @Override
+  public Failure exceptionToFailure(Throwable e) {
+    return this.encodeFailure(dataConverter.exceptionToFailure(e));
+  }
+
+  private Failure encodeFailure(Failure decodedFailure) {
+    Builder encodedFailure = decodedFailure.toBuilder();
+    if (decodedFailure.hasCause()) {
+      encodedFailure.setCause(encodeFailure(decodedFailure.getCause()));
+    }
+    if (decodedFailure.hasEncodedAttributes()) {
+      encodedFailure.setEncodedAttributes(encode(decodedFailure.getEncodedAttributes()));
+    }
+    switch (decodedFailure.getFailureInfoCase()) {
+      case APPLICATION_FAILURE_INFO:
+        {
+          ApplicationFailureInfo decodedInfo = decodedFailure.getApplicationFailureInfo();
+          ApplicationFailureInfo.Builder encodedInfo = decodedInfo.toBuilder();
+          if (decodedInfo.hasDetails()) {
+            encodedInfo.setDetails(encode(decodedInfo.getDetails()));
```

Comment:
@bergundy Can you please confirm that `Failure` details also have to be codec encoded, and that it is _non-optional_ (provided that codecs have indeed been registered)? That means that updated clients that are configured with codecs may suddenly be unable to decode failure details from history generated by previous workers.

Reply:
> That means that updated clients that are configured with codecs may suddenly be unable to decode failure details from history generated by previous workers.

That's not how Codecs should be implemented though. If implemented correctly. Codecs should check the `encoding` field and don't try to decode that they didn't encode in the first place.

Reply:
Very good point. Dismiss my argument.